File: Leetcode_JS/problem1518.js
function numWaterBottles(numBottles, numExchange) {
    let total = numBottles;
    while (numBottles>=numExchange) {
        fullB = Math.floor(numBottles/numExchange);
        total += fullB;
        numBottles = numBottles%numExchange+fullB;
    }
    return total;
}

// Test Cases
// console.log(numWaterBottles(1, 100));
// console.log(numWaterBottles(2, 2));
// console.log(numWaterBottles(3, 2));
// console.log(numWaterBottles(4, 2));
// console.log(numWaterBottles(5, 2));
// console.log(numWaterBottles(14, 4));
// console.log(numWaterBottles(25, 4));
// console.log(numWaterBottles(12, 3));
// console.log(numWaterBottles(9, 3));
// console.log(numWaterBottles(65, 3));

File: Leetcode_CPP/problem1518.cpp
#include <iostream>
#include <stdio.h>
#include <cmath>
using namespace std;

// time - O(log n)
// space - O(1)
int numWaterBottles(int numBottles, int numExchange)
{
    // initially numBottles will be drunked
    int count = numBottles; 

    while (numBottles >= numExchange)
    {
        // calculate the full bottles from total empty bottles which are numBottles/numExchange
        int fullB = numBottles / numExchange;
        count += fullB;

        // next numBottles(emptyBottles) will be remaining empty bottles + full Bottles
        numBottles = (numBottles%numExchange) + fullB;
    }
    
    return count;
}

int main()
{
    int res = numWaterBottles(12, 4);
    cout<<res<<endl;
    return 0;
}